### PR TITLE
New API for creating unbound machine ids and using them across runtime generations

### PR DESCRIPTION
**Pull request by pdeligia**

### CODE REVIEW

**Comment by akashlal** [comment]
I made a few modifications. I removed `CreateMachineId`. Further, `CreateMachine` will assert that the `MachineId` passed to it was never used before, not even for a machine that has now halted. Generations are retained but they have no role in bug-finding mode. Instead, the user can use `IncGenerationForTesting` to simulate the creation of `MachineId`s in different generations. I have checked in some tests. A full working sample is available here: `PSharpLab\Samples\Experimental\UnboundMachineIds`

**Comment by TedHartMS** [comment]
Looks good. It would be cleaner to factor out the new duplicated code in StateMachineRuntime/BugFindingRuntime.CreateMachine at the top (down to mid.Bind, and the MachineMap.TryAdd+assert) to common functions in the PSharpRuntime base class (MachineMap is already there).

File: Source/Core/Runtime/StateMachineRuntime.cs [hunk 1]
```diff
@@ -207,6 +237,17 @@ public override void RemoteSendEvent(MachineId target, Event e, Guid? operationG
...
 
         /// <summary>
+        /// Create a new machine Id with incremented generation. This
+        /// method should be used only for testing with PSharpTester.
+        /// </summary>
+        /// <param name="mid">Machine Id</param>
+        /// <returns>Machine Id with incremented generation count</returns>
+        public override MachineId IncGenerationForTesting(MachineId mid)
+        {
+            throw new NotImplementedException("The method IncGenerationForTesting should not be used for production runs.");
```
Comment (pdeligia) [inline Source/Core/Runtime/StateMachineRuntime.cs]: I think this should change to `NotSupportedException`, to make it 100% clear that we do not support this method in the production runtime (versus that it is not implemented now, but might be in the future).

**Comment by akashlal** [comment]
Thanks Pantazis. Can you please rebase? I can then take care of making these changes and testing.

Akash

**Comment by pdeligia** [comment]
Hi Akash, I just rebased and force pushed. Can you please delete your local branch, and checkout again as fresh, before working on this? (because of the rebase, else you will get weird conflicts if you pull).

Pantazis

**Comment by akashlal** [comment]
Thanks. How do I delete my local branch? Or should I do a fresh enlistment?

Akash

**Comment by TedHartMS** [comment]
https://stackoverflow.com/questions/9210446/replace-local-branch-with-remote-branch-entirely:  Make sure you're on the create-machineid branch then "git reset --hard origin/create-machineid". There are also steps in that link for doing it by deleting and re-fetching the local branch instead.

**Comment by akashlal** [comment]
I further applied the principle of "least number of bells and whistles". I found `IncGenerationForTesting` to be ugly and too specific to failover testing. I have instead re-introduced `CreateMachineId`. This makes the API more general. Further, I believe `CreateMachineId` is not a scheduling point for testing. 

@paulthomson @shazqadeer Do you guys agree that `CreateMachineId` (creates and returns a fresh machine-id) is not a scheduling point?

**Comment by paulthomson** [comment]
I think it does indeed *not* need to be a scheduling point. 

Out of interest, what is the reason for introducing CreateMachineId? How is it used?

**Pull request merged**